### PR TITLE
Implement auto sign in AB test

### PR DESCRIPTION
**Pull request by Mullefa**
When merged, 50% of payment failure emails will include an auto sign-in token; the other 50% will include no token. 

The queries used to derive the data from this test will be similar to the [queries](https://github.com/guardian/identity-platform/blob/master/docs/08-data-lake/email-token-ab-test.sql) used for the encrypted email token test #22 .

### CODE REVIEW

File: payment-failure/src/main/scala/com/gu/identity/paymentfailure/BrazeEmailService.scala [hunk 1]
```diff
@@ -72,16 +72,16 @@ class DefaultBrazeEmailService(
...
 }
 
-// (modulo errors) half of the emails sent will include the encrypted email token;
-// the other half will include no sign-in tokens.
-// There will be additional meta data (abName and abVariant) to facilitate tracking which segment a user is in.
-class BrazeEmailServiceWithEncryptedBrazeEmailTest(
+// Used to send an email with additional trigger properties (aka Braze metadata) derived from a variant in an AB test.
+// For example abName, abVariant and an additional token to facilitate sign-in.
+// See e.g. EncryptedEmailTest for a concrete example.
+class BrazeEmailServiceWithAbTest(
```
Comment (Mullefa) [inline payment-failure/src/main/scala/com/gu/identity/paymentfailure/BrazeEmailService.scala:78]: This has been made slightly more generic so an email can be sent with an arbitrary AB test.

File: payment-failure/src/main/scala/com/gu/identity/paymentfailure/abtest/AutoSignInTest.scala [hunk 2]
```diff
@@ -0,0 +1,40 @@
+package com.gu.identity.paymentfailure.abtest
+
+import cats.syntax.either._
+import com.gu.identity.paymentfailure.{BrazeClient, IdentityClient}
+import com.gu.identity.paymentfailure.IdentityClient.AutoSignInLinkRequestBody
+
+import scala.util.Random
+
+class AutoSignInTest(identityClient: IdentityClient) extends VariantGenerator {
```
Comment (Mullefa) [inline payment-failure/src/main/scala/com/gu/identity/paymentfailure/abtest/AutoSignInTest.scala:7]: New auto sign-in test

File: payment-failure/src/main/scala/com/gu/identity/paymentfailure/abtest/EncryptedEmailTest.scala [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package com.gu.identity.paymentfailure.abtest
+
+import cats.syntax.either._
+import com.gu.identity.paymentfailure.{BrazeClient, IdentityClient}
+import com.gu.identity.paymentfailure.IdentityClient.IdentityEmailTokenRequest
+
+import scala.util.Random
+
+class EncryptedEmailTest(identityClient: IdentityClient) extends VariantGenerator {
```
Comment (Mullefa) [inline payment-failure/src/main/scala/com/gu/identity/paymentfailure/abtest/EncryptedEmailTest.scala:9]: Previous test, refactored to implement the `VariantGenerator` trait. For now, keeping this file so we have a record-in-code of the tests run.

File: payment-failure/src/main/scala/com/gu/identity/paymentfailure/abtest/VariantGenerator.scala [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package com.gu.identity.paymentfailure.abtest
+
+// metadata in the variant is included in the metadata sent to Braze.
+// See BrazeEmailServiceWithAbTest for more context.
+case class Variant(testName: String, variantName: String, metadata: Map[String, String] = Map.empty)
+
+// Used to generate a variant for a test.
+trait VariantGenerator {
```
Comment (Mullefa) [inline payment-failure/src/main/scala/com/gu/identity/paymentfailure/abtest/VariantGenerator.scala:10]: Used to make the AB test email service more generic.

**Review by laurenemms — APPROVED** [review]

**Review by laurenemms — APPROVED** [review]

**Pull request merged**